Cut dependencies on future modules

lrytz opened the pull request:
Re-submission of #5677

adriaanm committed:
Avoid `sys.props`

adriaanm committed:
Avoid `sys.error`

Use `throw new RuntimeException` (or a subclass) instead.

adriaanm committed:
Avoid `sys.exit`

adriaanm committed:
Avoid `sys.process`

lrytz commented:
My changes compared to #5677 

- Removed new deprecations
- Rebased

Note that https://github.com/scala/scala/commit/6411170ede09ace8e5f1f018dcf5661846ed7d27 recently added a new method `scala.util.Properties.coloredOutputEnabled`, which I left untouched.

lrytz committed:
Avoid using code that will move to modules

`scala.concurrent`, `scala.Console`, `scala.compat._`,
`sys.runtime`, `sys.addShutdownHook`, `sys.allThreads`,
`sys.env`.

Also fixes a test reducing references to static modules (Console, Predef)
delays initialization of object `NoStackTrace`. In `t2318`, this may now
happen *after* the security manager is installed. Therefore the security
manager needs to allow reading the system property
`scala.control.noTraceSuppression`

lrytz commented:
The remaining test failures look like this:

```
[error] Test tools.test.osgi.reflection.basic.BasicReflectionTest.basicMirrorThroughOsgi failed: scala.reflect.internal.MissingRequirementError: package object scala.sys not found., took 1.455 sec
[error]     at scala.reflect.internal.MissingRequirementError$.signal(MissingRequirementError.scala:17)
[error]     at scala.reflect.internal.MissingRequirementError$.notFound(MissingRequirementError.scala:18)
[error]     at scala.reflect.internal.Mirrors$RootsBase.getPackageObject(Mirrors.scala:184)
[error]     at scala.reflect.internal.Mirrors$RootsBase.getPackageObject(Mirrors.scala:180)
[error]     at scala.reflect.internal.Definitions$DefinitionsClass.SysPackage$lzycompute(Definitions.scala:339)
[error]     at scala.reflect.internal.Definitions$DefinitionsClass.SysPackage(Definitions.scala:339)
[error]     at scala.reflect.runtime.JavaUniverseForce.force(JavaUniverseForce.scala:244)
[error]     at scala.reflect.runtime.JavaUniverseForce.force$(JavaUniverseForce.scala:6)
[error]     at scala.reflect.runtime.JavaUniverse.force(JavaUniverse.scala:16)
[error]     at scala.reflect.runtime.JavaUniverse.init(JavaUniverse.scala:147)
[error]     at scala.reflect.runtime.JavaUniverse.<init>(JavaUniverse.scala:78)
[error]     at scala.reflect.runtime.package$.universe$lzycompute(package.scala:17)
[error]     at scala.reflect.runtime.package$.universe(package.scala:17)
[error]     at tools.test.osgi.reflection.basic.BasicReflectionTest.basicMirrorThroughOsgi(BasicReflection.scala:56)
```

The osgi tests all pass if this access to `sys.props` (removed in this PR) stays in `SymbolTable`:

```diff
--- a/src/reflect/scala/reflect/internal/SymbolTable.scala
+++ b/src/reflect/scala/reflect/internal/SymbolTable.scala
@@ -138,7 +138,8 @@ abstract class SymbolTable extends macros.Universe

   /** Dump each symbol to stdout after shutdown.
    */
-  final val traceSymbolActivity = System.getProperty("scalac.debug.syms") != null
+  final val traceSymbolActivity = sys.props contains "scalac.debug.syms"
   object traceSymbols extends {
     val global: SymbolTable.this.type = SymbolTable.this
   } with util.TraceSymbolActivity
```

lrytz commented:
I'm looking into this one, I think I'm getting somewhere.

adriaanm commented:
Thanks! Re: osgi, does this have something to do with osgi enforcing boundaries and somehow the dependency on scala.sys was inferred and not explicitly specified?

lrytz commented:
No, I don't think it has to do with osgi itself, but with the classloader type used in runtime reflection. It seems to be specific to package objects. I'll figure it out I think.

adriaanm commented:
> I'll figure it out I think.

Na klar!

lrytz commented:
Finally figured it out... Your thorough work removed each and every dependency from scala-reflect to `scala.sys`, which makes the `scala.sys` package completely disappear from the `META-INF/MANIFEST.MF` in `scala-reflect.jar`. Adding any reference to `scala.sys` within the reflect codebase fixes it. For example i added `class FlupediFlup { val p = scala.sys.Prop }` to `scala.reflect.io`, then the manifests gets

```
...
Export-Package: ...
  scala.reflect.io ...uses:=" ... ,scala.sys"
...
Import-Package: ..., scala.sys ...
```

lrytz committed:
Replace mkSysErrorCall by mkThrowNewRuntimeException

Since we no longer want to depend on `sys.error`, the compiler
also shouldn't generate calls to it.

This also fixes the OSGI tests. The failure was interesting: once
there were no more refernces to `scala.sys` left in scala-reflect,
the dependency on that package disappeared from scala-reflect.jar's
MANIFEST.MF. When instantiating scala.reflect.runtime.universe through
an osgi classloader, `getPackageObject("scala.sys")` in definitions
failed becasue the package could not be found through that classloader.

lrytz commented:
Now this is really weird: the `MainGenericRunner` (which calls `System.exit`) no longer waits for non-daemon threads to finish after this PR.. I haven't found the reason yet.

```scala
$ cat Test.scala
object Test {
  def main(args: Array[String]) {
    new Thread(new Runnable {
      def run = { Thread.sleep(1000); println("hi") }
    }).start
  }
}
$ scalac12 Test.scala
$ time scala12 Test.scala
hi
/Users/luc/scala/scala-2.12/bin/scala Test.scala  0.77s user 0.09s system 42% cpu 2.042 total
$ time ../build/quick/bin/scala Test
../build/quick/bin/scala Test  0.68s user 0.13s system 103% cpu 0.780 total
```

lrytz commented:
Oh, not that weird after all.. https://github.com/scala/scala/pull/5830/files#diff-815852a7f5eda790669df133ce921662L104

lrytz commented on src/repl/scala/tools/nsc/MainGenericRunner.scala:
```diff
@@ -101,5 +101,5 @@ class MainGenericRunner {
 
 object MainGenericRunner extends MainGenericRunner {
-  def main(args: Array[String]): Unit = if (!process(args)) sys.exit(1)
+  def main(args: Array[String]): Unit = System.exit(if (process(args)) 0 else 1)
```
This changes behavior.. if `process` is true, we should not call `exit`. Otherwise this will kill the application without waiting for spawned non-daemon threads to finish. https://github.com/scala/scala/blob/v2.12.1/test/files/run/lazy-concurrent.scala is an example.

adriaanm replied:
Ouch! My bad :-/

lrytz commented:
Green now!

lrytz commented:
I'll squash some of the smaller commits into one to make them fewer.
